mcode/mcode-extraction-framework
# Template revamp

# Summary
Begins the revamp of our template approach, backing away from EJS and moving towards home-rolled JavaScript template functions that receive template data and produce JSONified FHIR resources. This task starts with the Patient resource, adding fundamental snippets and decorators for the rapid and reusable creation of future templates. Additionally, the approach taken here is one that aims to be progressive; the previous approach for templating with EJS is still used for non-updated templates. As we update additional templates, we need to make minor changes to our `EJSUtils` file until we are finished with the EJS approach, at which point we'll need to remove that file and update the `templateUtils` accordingly. 

# Detailed Notes
Below is a list of detailed notes on the 
- `EJSUtils` has a list of updated templates, as well as an update to the FHIRTemplateLookup table to point at the newly created Patient TemplateFunction. 
- `templateUtils` (where some of the old `EJSUtils` functionality will eventually live) contains several decorators for templateFunctions. These decorators enable the separation of requirements for which arguments need to be present from the nature of the template itself. 
- Various places use the below motif for adding a property and value only-if the provided value is non-falsy. This is helpful in avoiding empty properties on the resultant JSON object, though there are other ways we could trim these properties if we wanted to explore them. Note also that `...null` evaluates without throwing an error, and produces no new property.
```javascript
{
    address: [
      {
        ...(addressLine && {
          line: [
            addressLine,
          ],
        }),
        ...(city && { city }),
        ...(state && { state }),
        ...(zip && { postalCode: zip }),
        country: 'US',
      },
    ],
  };
```

# Making New Templates
Making a new template involves the following steps: 
- Create a new `.js` file responsible for taking incoming data as described by the current `.ejs` template and producing the JSONified FHIR Resources.
- Uses the decorators, `...(property && {property})` pattern and any number of smaller functions you create to render the various subcomponents of the Patient resource, 
- Add tests for any new sub-functions, and update previous template tests to ensure that it tests the new templateFunction
- Update the `EJSUtils` file to ensure that the `mcodeProfileID` is added to the new templates list, ensuring that the render function uses the appropriate render logic for that template. 

# Testing guidance
- Ensure all tests are passing; 
- Ensure that there aren't any other tests that need updating;
- Review the current snippets and comment on any ones we need to add;

Happy Reviewing!
